modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/mv/RowVersion.java

```diff
@@ -0,0 +1,137 @@
+import org.apache.ignite.internal.pagememory.io.AbstractDataPageIo;
+import org.apache.ignite.internal.pagememory.io.IoVersions;
+import org.apache.ignite.internal.storage.pagememory.mv.io.RowVersionDataIo;
+import org.apache.ignite.internal.tx.Timestamp;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Represents row version inside row version chain.
+ */
+public class RowVersion extends StorableBase {
+    /**
+     * A 'timestamp' representing absense of a timestamp.
+     */
+    public static final Timestamp NULL_TIMESTAMP = new Timestamp(Long.MIN_VALUE, Long.MIN_VALUE);
+    /**
+     * Represents an absent partitionless link.
+     */
+    public static final long NULL_LINK = 0;
+
+    @Nullable
+    private final Timestamp timestamp;
+    private final long nextLink;
+    private final int valueSize;
+    private final ByteBuffer value;
+
+    /**
+     * Constructor.
+     */
+    public RowVersion(int partitionId, long nextLink, ByteBuffer value) {
+        this(partitionId, 0, null, nextLink, value);
+    }
+
+    /**
+     * Constructor.
+     */
+    public RowVersion(int partitionId, long link, @Nullable Timestamp timestamp, long nextLink, ByteBuffer value) {
+        super(partitionId);
+        link(link);
+
+        assert !NULL_TIMESTAMP.equals(timestamp) : "Null timestamp provided";
+
+        this.timestamp = timestamp;
+        this.nextLink = nextLink;
+        this.valueSize = value.limit();
+        this.value = value;
+    }
+
+    @Nullable
+    public Timestamp timestamp() {
+        return timestamp;
+    }
+
+    public Timestamp timestampForStorage() {
+        return timestamp == null ? NULL_TIMESTAMP : timestamp;
+    }
+
+    /**
+     * Returns partitionless link of the next version or {@code 0} if this version is the last in the chain (i.e. it's the oldest version).
+     *
+     * @return partitionless link of the next version or {@code 0} if this version is the last in the chain
+     */
+    public long nextLink() {
+        return nextLink;
+    }
+
+    public int valueSize() {
+        return valueSize;
+    }
+
+    public ByteBuffer value() {
+        return value;
+    }
+
+    public boolean hasNextLink() {
+        return nextLink != NULL_LINK;
+    }
+
+    boolean isTombstone() {
+        return valueSize() == 0;
+    }
+
+    boolean isUncommitted() {
+        return timestamp == null;
+    }
+
+    boolean isCommitted() {
+        return timestamp != null;
+    }
+
+    @Override
+    public int size() {
+        return timestampStoreSize() + nextLinkStoreSize() + valueStoreSize();
+    }
+
+    private int timestampStoreSize() {
+        return 2 * Long.BYTES;
+    }
+
+    private int nextLinkStoreSize() {
+        return PartitionlessLinks.PARTITIONLESS_LINK_SIZE_BYTES;
+    }
+
+    private int valueStoreSize() {
+        return Integer.BYTES + value.limit();
+    }
+
+    @Override
+    public int headerSize() {
+        return 0;
```

Comment:
I specifically mentioned header sizes in JIRA. Please put "size()" here. This is a bug right now

Reply:
`headerSize()` is only used for fragmented case to make sure that the 'header' is not torn apart and it always remains on the first page. Why is it a bug if we return 0 here?

Please note that in Ignite 2, the only implementation that returns something non-zero from this method is `MvccDataRow`. Other implementations either return 0, or just throw an exception.

Reply:
After a conversation, it was decided that a meaningful header size is better as it makes reading code simpler. I implemented the method as suggested.